src/features/editors/VisualHTMLEditor.tsx

```diff
@@ -0,0 +1,92 @@
+import './VisualHTMLEditor.css'
+import {CodeSegment, CodeSegments} from "../../editorconfig"
+import Line from "./VisualHTMLEditorLine"
+import {DragDropContext, DropResult} from 'react-beautiful-dnd'
+import {useAppDispatch} from "../../hooks"
+import {moveElement} from "../../store/features/editors/visualHTML";
+
+interface IVisualHTMLEditor {
+  elements: CodeSegments
+}
+
+export type LineSegment = {
+  id: string,
+  type: string,
+  value: string,
+  unlocked: boolean,
+  index: number
+}
+
+export type LineSegments = Array<LineSegment>
+
+type TLine = {
+  lineContents: LineSegments,
+  lineIndentation: number
+}
+
+function VisualHTMLEditor(props: IVisualHTMLEditor) {
+  let indentCounter = 0
+  let lineBuilder: LineSegments = []
+  let lines: Array<TLine> = []
+  let prevWasBlock: boolean = true
+
+  const dispatch = useAppDispatch()
+
+  const indenter = (element: CodeSegment) => element.type === 'opening' ? indentCounter++ : indentCounter--
```

Comment:
I may have forgotten this in the previous iteration, but this type should be an enum: it seems to have a delimited and known set of values.

Reply:
I see! If I understand correctly, when a type has a set of know values it should be defined as an enum?
Here's my change:
```
enum TagType {
  Opening = 'opening',
  Closing = 'closing',
  Text = 'text'

}

export interface CodeSegment {
  id: string,
  type: TagType,
  value: string,
  unlocked: boolean
}
```
Is this the correct way of doing it?
Thanks :)

Reply:
Exactly. If a variable value is defined by the user, we can simply say that its type is string (or something else, depending of the value). But if the variable value is defined by the code, we know by advance the set of values that it can take. So it should be an enum by definition (enumerated set of values), just like how you've created it :+1:

Reply:
Cool, understood! Thanks again